[%intel] deprecation message

rscohn2 opened the pull request:
Add a method to deprecate a toolchain and use it to deprecate %intel and %dpcpp. You can choose between a warning and an error. Using the `--deprecated` option disables the warning/error. The check is at the point the build environment is set up so it is emitted once for  each package that is built with a deprecated toolchain.

Edit: The comment below describes an earlier implementation that was abandoned.
See the description below in the comments. After seeing how it works, I am no longer sure spack should do anything. It might be annoying to see the warning if you are not using the `%intel` compiler. Having the compiler wrapper emit a warning is unnecessary because the compiler already emits a warning.

```
rscohn1@gkdse-fcp-19:spack$ spack install patchelf
==> Warning: The %intel toolchain has been deprecated. Remove %intel from your compilers.yaml and use %oneapi instead.
^[]0;Spack: Processing gmake [1/2]^G[+] /usr (external gmake-4.3-5r2svr4ltlqsp3ue4utnbipfmeigph6s)
^[]0;Spack: Processing patchelf [2/2]^G^[]0;Spack: Acquiring lock for patchelf [2/2]^G==> Waiting for patchelf-0.17.2-p6q7rn4hwm62ua\dx5tobh4ot7kqkinfy
```

rscohn2 committed:
[%intel] deprecation message

wyphan commented:
Yeah, I agree that the Spack warning is redundant, and potentially confusing when it comes from here:

>       # The warning is triggered by having an intel toolchain in
>       # your compilers.yaml, even if you do not use it.

Note that some facilities might keep an older Intel toolchain around tho, just because they already paid for support :joy:

rscohn2 committed:
warn on use of icc/dpcpp

rscohn2 committed:
fix style

rscohn2 commented:
@wyphan: I moved the check to a place where the compiler is actually used, not just configured. Now you will not get a message for adding %intel to the compilers.yaml. It is still a warning, and I added support for making it an error at some point in the future. It can be disabled with --deprecated. Also deprecated %dpcpp toolchain
```
(venv) rscohn1@gkdse-fcp-19:spack$ spack uninstall --all -y patchelf ; spack install patchelf%dpcpp
==> Error: patchelf does not match any installed packages.
^[]0;Spack: Processing gmake [1/2]^G[+] /usr (external gmake-4.3-zal5fnmd7xcbk6lhplxn62fdedlidtqt)
^[]0;Spack: Processing patchelf [2/2]^G^[]0;Spack: Acquiring lock for patchelf [2/2]^G==> Waiting for patchelf-0.17.2-gwsv7wkh6cwiba\
vh5zz7bcz7csxveulv
^[]0;Spack: Preparing patchelf [2/2]^G^[]0;Spack: Installing patchelf [2/2]^G==> Installing patchelf-0.17.2-gwsv7wkh6cwibavh5zz7bcz7\
csxveulv [2/2]
==> No binary for patchelf-0.17.2-gwsv7wkh6cwibavh5zz7bcz7csxveulv found: installing from source
==> Warning: The %dpcpp toolchain has been deprecated, use %oneapi with -fsycl.
```

alalazo commented:
@rscohn2 `dpcpp` has been removed, and https://github.com/spack/spack/pull/43805 changed the default compiler preferences to account for `%oneapi`, and removed `%intel`. Would that be sufficient to close this PR? Or is there something missing?

rscohn2 commented:
> Would that be sufficient to close this PR? Or is there something missing?

Hi @alalazo: I was concerned about people not knowing that new oneapi releases no longer include %intel toolchain. Making %intel builds error out would be too disruptive and so I only emitted a message. However, it's not clear that anyone would see the message if the build succeeded. I haven't heard any complaints from users of the new releases so maybe people are aware and the message is unnecessary. Do you think any extra steps are needed?

alalazo commented:
@rscohn2 I made a change in #43778 that prevents detecting `%intel` in OneAPI 2024. If that was the issue, I think we can close this PR. Let me know if you agree.

rscohn2 commented:
It's ok, closing.
